File: backend/app/schemas/music.py
from typing import List, Optional, Any, Dict
from pydantic import BaseModel, Field


class Thumbnail(BaseModel):
    url: str
    width: Optional[int] = None
    height: Optional[int] = None


class ArtistBasic(BaseModel):
    id: Optional[str] = None
    name: str


class Track(BaseModel):
    videoId: str
    title: str
    artists: List[ArtistBasic] = Field(default_factory=list)
    album: Optional[str] = None
    albumId: Optional[str] = None
    duration: Optional[str] = "0:00"
    duration_seconds: Optional[int] = 0
    thumbnails: List[Thumbnail] = Field(default_factory=list)
    thumbnail: Optional[str] = ""
    isExplicit: Optional[bool] = False


class AlbumBasic(BaseModel):
    browseId: str
    title: str
    type: Optional[str] = "Album"
    year: Optional[str] = None
    artists: List[ArtistBasic] = Field(default_factory=list)
    thumbnails: List[Thumbnail] = Field(default_factory=list)
    thumbnail: Optional[str] = ""
    trackCount: Optional[int] = None


class Artist(BaseModel):
    channelId: str
    name: str
    description: Optional[str] = None
    subscribers: Optional[str] = None
    views: Optional[str] = None
    thumbnails: List[Thumbnail] = Field(default_factory=list)
    thumbnail: Optional[str] = ""
    topSongs: List[Track] = Field(default_factory=list)
    albums: List[AlbumBasic] = Field(default_factory=list)
    singles: List[AlbumBasic] = Field(default_factory=list)
    relatedArtists: List[ArtistBasic] = Field(default_factory=list)


class Album(BaseModel):
    browseId: str
    title: str
    description: Optional[str] = None
    artists: List[ArtistBasic] = Field(default_factory=list)
    year: Optional[str] = None
    trackCount: Optional[int] = 0
    duration: Optional[str] = None
    thumbnails: List[Thumbnail] = Field(default_factory=list)
    thumbnail: Optional[str] = ""
    tracks: List[Track] = Field(default_factory=list)


class PlaylistSummary(BaseModel):
    id: str
    title: str
    description: Optional[str] = ""
    author: Optional[str] = ""
    itemCount: Optional[int] = 0
    thumbnails: List[Thumbnail] = Field(default_factory=list)
    thumbnail: Optional[str] = ""


class ShelfItem(BaseModel):
    id: str
    title: str
    subtitle: Optional[str] = ""
    type: str  # "song", "album", "artist", "playlist"
    thumbnail: Optional[str] = ""
    thumbnails: List[Thumbnail] = Field(default_factory=list)
    artists: List[ArtistBasic] = Field(default_factory=list)
    videoId: Optional[str] = None
    browseId: Optional[str] = None
    duration: Optional[str] = None


class Shelf(BaseModel):
    title: str
    contents: List[ShelfItem] = Field(default_factory=list)


class HomeResponse(BaseModel):
    shelves: List[Shelf] = Field(default_factory=list)


class SearchResults(BaseModel):
    query: str
    topResult: Optional[Dict[str, Any]] = None
    songs: List[Track] = Field(default_factory=list)
    albums: List[AlbumBasic] = Field(default_factory=list)
    artists: List[Dict[str, Any]] = Field(default_factory=list)
    playlists: List[PlaylistSummary] = Field(default_factory=list)
    videos: List[Track] = Field(default_factory=list)


class LyricWord(BaseModel):
    text: str
    startTime: int  # milliseconds
    endTime: Optional[int] = None  # milliseconds


class LyricLine(BaseModel):
    id: str
    text: str
    startTime: Optional[int] = None  # milliseconds
    endTime: Optional[int] = None  # milliseconds
    words: Optional[List[LyricWord]] = None
    script: Optional[str] = None  # e.g., "arabic", "devanagari", "bengali", "gurmukhi", "latin", "cjk", etc.
    direction: Optional[str] = None  # "ltr", "rtl", "auto"
    language: Optional[str] = None  # e.g., "ur", "ar", "fa", "hi", "bn", "pa", "en", "unknown"
    translation: Optional[str] = None
    romanization: Optional[str] = None


class LyricsResponse(BaseModel):
    videoId: str
    synced: bool = False
    hasLyrics: bool = False
    lines: List[LyricLine] = Field(default_factory=list)
    lyrics: Optional[str] = None
    source: Optional[str] = None
    provider: Optional[str] = None
    syncConfidence: Optional[str] = None
    lyricsDuration: Optional[float] = None
    mediaDuration: Optional[float] = None
    durationDifference: Optional[float] = None
    primaryScript: Optional[str] = None
    direction: Optional[str] = None  # "ltr", "rtl", "auto"
    language: Optional[str] = None
    instrumental: Optional[bool] = False

File: frontend/src/types/music.ts
export interface Thumbnail {
  url: string;
  width?: number;
  height?: number;
}

export interface ArtistBasic {
  id?: string | null;
  name: string;
}

export interface Track {
  videoId: string;
  title: string;
  artists: ArtistBasic[];
  album?: string | null;
  albumId?: string | null;
  duration?: string;
  duration_seconds?: number;
  thumbnails?: Thumbnail[];
  thumbnail?: string;
  isExplicit?: boolean;
}

export interface AlbumBasic {
  browseId: string;
  title: string;
  type?: string;
  year?: string | null;
  artists?: ArtistBasic[];
  thumbnails?: Thumbnail[];
  thumbnail?: string;
  trackCount?: number | null;
}

export interface Album {
  browseId: string;
  title: string;
  description?: string;
  artists: ArtistBasic[];
  year?: string | null;
  trackCount?: number;
  duration?: string;
  thumbnails?: Thumbnail[];
  thumbnail?: string;
  tracks: Track[];
}

export interface Artist {
  channelId: string;
  name: string;
  description?: string;
  subscribers?: string;
  views?: string;
  thumbnails?: Thumbnail[];
  thumbnail?: string;
  topSongs: Track[];
  albums: AlbumBasic[];
  singles: AlbumBasic[];
  relatedArtists: ArtistBasic[];
}

export interface FollowedArtist {
  id: string;
  name: string;
  thumbnail?: string;
  subscribers?: string;
  songs?: string;
}

export interface PlaylistSummary {
  id: string;
  title: string;
  description?: string;
  author?: string;
  itemCount?: number;
  thumbnails?: Thumbnail[];
  thumbnail?: string;
}

export interface CustomPlaylistTrack {
  id: string;
  playlist_id: string;
  video_id: string;
  title: string;
  artist: string;
  album?: string;
  thumbnail_url?: string;
  duration?: string;
  duration_seconds?: number;
  position: number;
  added_at?: string;
}

export interface CustomPlaylist {
  id: string;
  title: string;
  description?: string;
  thumbnail_url?: string;
  track_count: number;
  created_at?: string;
  updated_at?: string;
  tracks?: CustomPlaylistTrack[];
}

export interface ShelfItem {
  id: string;
  title: string;
  subtitle?: string;
  type: 'song' | 'album' | 'artist' | 'playlist';
  thumbnail?: string;
  thumbnails?: Thumbnail[];
  artists?: ArtistBasic[];
  videoId?: string;
  browseId?: string;
  duration?: string;
}

export interface Shelf {
  title: string;
  contents: ShelfItem[];
}

export interface HomeResponse {
  shelves: Shelf[];
}

export interface SearchResults {
  query: string;
  topResult?: {
    type: string;
    data: any;
  } | null;
  songs: Track[];
  albums: AlbumBasic[];
  artists: {
    channelId: string;
    name: string;
    subscribers?: string;
    thumbnail?: string;
  }[];
  playlists: PlaylistSummary[];
  videos: Track[];
}

export interface LyricWord {
  text: string;
  startTime: number;
  endTime?: number;
}

export interface LyricLine {
  id: string;
  text: string;
  startTime?: number;
  endTime?: number;
  words?: LyricWord[];
  script?: string;
  direction?: 'ltr' | 'rtl' | 'auto';
  language?: string;
  translation?: string;
  romanization?: string;
}

export interface LyricsResponse {
  videoId: string;
  synced: boolean;
  hasLyrics: boolean;
  lines: LyricLine[];
  lyrics?: string | null;
  source?: string | null;
  provider?: 'youtube_musixmatch' | 'open_synced_lrclib' | 'lrclib' | string | null;
  syncConfidence?: 'excellent' | 'good' | 'uncertain' | 'poor' | null;
  lyricsDuration?: number | null;
  mediaDuration?: number | null;
  durationDifference?: number | null;
  primaryScript?: string | null;
  direction?: 'ltr' | 'rtl' | 'auto' | null;
  language?: string | null;
  instrumental?: boolean;
}

export type RepeatMode = 'off' | 'queue' | 'track';

export interface PlaybackHistoryItem {
  id: string;
  video_id: string;
  title: string;
  artist: string;
  album?: string;
  thumbnail_url?: string;
  duration?: string;
  played_at?: string;
  play_count: number;
}
